curp/src/client_new/unary.rs

```diff
@@ -0,0 +1,620 @@
+        if !linearizable {
+            // firstly, try to fetch the local server
+            if let Some(connect) = <Unary<C> as ClientApi>::local_connect(self) {
+                /// local timeout, in fact, local connect should only be bypassed, so the timeout maybe unused.
+                const FETCH_LOCAL_TIMEOUT: Duration = Duration::from_secs(1);
+
+                let resp = connect
+                    .fetch_cluster(FetchClusterRequest::default(), FETCH_LOCAL_TIMEOUT)
+                    .await
+                    .unwrap_or_else(|e| {
+                        unreachable!(
+                            "fetch cluster from local connect should never failed, err {e:?}"
+                        )
+                    })
+                    .into_inner();
+                return Ok(resp);
+            }
+        }
+        // then fetch the whole cluster
+        let (quorum, mut responses) = self.for_each_server(|conn| async move {
+            (
+                conn.id(),
+                conn.fetch_cluster(FetchClusterRequest { linearizable }, timeout)
+                    .await
+                    .map(Response::into_inner),
+            )
+        });
+        let majority_quorum = majority_quorum(quorum);
+
+        let mut max_term = 0;
+        let mut res = None;
+        let mut ok_cnt = 0;
+        let mut err = None;
+
+        while let Some((id, resp)) = responses.next().await {
+            let inner = match resp {
+                Ok(r) => r,
+                Err(e) => {
+                    warn!("fetch cluster from {} failed, {:?}", id, e);
+                    if e.return_early() {
+                        return Err(e);
+                    }
+                    err = Some(e);
+                    continue;
+                }
+            };
+
+            #[allow(clippy::integer_arithmetic)]
+            match max_term.cmp(&inner.term) {
+                Ordering::Less => {
+                    if !inner.members.is_empty() {
+                        max_term = inner.term;
+                        res = Some(inner);
+                    }
+                    ok_cnt = 1;
+                }
+                Ordering::Equal => {
+                    if !inner.members.is_empty() {
+                        res = Some(inner);
+                    }
+                    ok_cnt += 1;
+                }
+                Ordering::Greater => {}
+            }
+
+            if ok_cnt >= majority_quorum {
+                break;
+            }
+        }
+        if let Some(res) = res {
+            debug!("fetch cluster succeeded, result: {res:?}");
+            if let Err(e) = self
+                .cluster_state
+                .check_and_update(res.cluster_version, res.clone().into_members_addrs())
+                .await
+            {
+                warn!("update to a new cluster state failed, error {e}");
+            } else {
+                // FIXME: The updates of leader and cluster states are not atomic,
```

Comment:
How come? I don't get the point.

Reply:
Assuming that two `FetchClusterRequests` invoke simultaneously, when obtaining the results, we first update the cluster state and then update the leader state; this is not atomic. The final result may be that we obtain the cluster state of the second `FetchClusterResponse` and the leader state of the first `FetchClusterResponse`.

Reply:
Now it is atomic